plugins/rapid7_insight_agent/icon_rapid7_insight_agent/util/graphql_api/api_connection.py

```diff
@@ -316,42 +334,55 @@ def _get_agent(self, agent_input: str, agent_type: str) -> dict:
-            "variables": {"orgId": self.org_key},
-        }
-
-        self.logger.info("Getting all agents...")
-        results_object = self._post_payload(payload)
-
-        has_next_page = (
-            results_object.get("data", {})
-            .get("organization", {})
-            .get("assets", {})
-            .get("pageInfo", {})
-            .get("hasNextPage")
-        )
-        agents.extend(self._get_agents_from_result_object(results_object))
-        self.logger.info("Initial agents received.")
-
-        agent = self._find_agent_in_agents(agents, agent_input, agent_type)
-        if agent:
-            return agent
-
-        # See if we have more pages of data, if so get next page and append until we reach the end
-        self.logger.info(f"Extra pages of agents: {has_next_page}")
-        while has_next_page:
+        has_next_page = True
+        if not next_cursor:
+            payload = {
+                "query": "query( $orgId:String! ) { organization(id: $orgId) { assets( first: 10000 ) { pageInfo { hasNextPage endCursor } edges { node { id platform host { vendor version description hostNames { name } primaryAddress { ip mac } uniqueIdentity { source id } attributes { key value } } publicIpAddress location { city region countryName countryCode continent } agent { agentSemanticVersion agentStatus quarantineState { currentState } } } } } } }",
+                "variables": {"orgId": self.org_key},
+            }
+            self.logger.info("Getting first page of agents...")
+            results_object = self._post_payload(payload)
+
+            agents.extend(self._get_agents_from_result_object(results_object))
+            self.logger.info("Initial agents received.")
+
+            agent = self._find_agent_in_agents(agents, agent_input, agent_type)
+            # See if we have more pages of data, if so get next page and append until we reach the end
+            has_next_page = (
+                results_object.get("data", {})
+                .get("organization", {})
+                .get("assets", {})
+                .get("pageInfo", {})
+                .get("hasNextPage")
+            )
+            next_cursor = (
+                results_object.get("data", {})
+                .get("organization", {})
+                .get("assets", {})
+                .get("pageInfo", {})
+                .get("endCursor")
+            )
+        while agent is None and has_next_page:
+            current_time = datetime.now()
+            if (current_time - start_time) > timedelta(minutes=3):
```

Comment:
should we set the timeout as a variable at the top of this file so we can easily update it later and it gets applied to both paging API calls? Also wondering what the timeout is, should we drop this to 2 minutes as it still needs to filter on the results gathered so far?

Reply:
The filtering seems to take very little time based on cloud logs. We timeout at 4 minutes so to not potentially disrupt existing workflows I kept this quite high at 3 minutes.